scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,416 @@
+    def dimension(self):
+        return self.nodes.shape[-1]
+
+    # returns integral_estimate, error_estimate
+    def estimate(self, f, a, b):
+        """
+        Calculate estimate of integral of f in region described by corners a and b.
+
+        f should accept arrays of shape (input_dim, num_eval_points) and return arrays
+        of shape (output_dim_1, ..., output_dim_n, num_eval_points).
+        """
+        raise NotImplementedError
+
+
+class CubatureRuleLinearError(CubatureRule):
+    def __init__(self, nodes, weights, error_nodes, error_weights):
+        super().__init__(nodes, weights)
+
+        self.error_nodes = error_nodes
+        self.error_weights = error_weights
+
+    def estimate(self, f, a, b):
+        # TODO: need to output some sort of error when the rule and f have inconsistent
+        # dimensions
+
+        # The underlying cubature rule is presumed to be for the hypercube [0, 1]^n.
+        #
+        # To handle arbitrary regions of integration, it's necessary to apply a linear
+        # change of coordinates to map each interval [a[i], b[i]] to [-1, 1].
+        a = a[:, np.newaxis]
+        b = b[:, np.newaxis]
+
+        lengths = b - a
+
+        # nodes have shape (input_dim, eval_points)
+        nodes = (self.nodes + 1) * (lengths / 2) + a
+        error_nodes = (self.error_nodes + 1) * (lengths / 2) + a
+
+        # Also need to multiply the weights by a scale factor equal to the determinant
+        # of the Jacobian for this coordinate change.
+        weight_scale_factor = math.prod(lengths / 2)
+
+        # weights have shape (eval_points,)
+        weights = self.weights * weight_scale_factor
+        error_weights = self.error_weights * weight_scale_factor
+
+        # f(nodes) will have shape (eval_points, output_dim)
+        # integral_estimate should have shape (output_dim)
+        integral_estimate = np.sum(weights * f(nodes), axis=-1)
+
+        # print(integral_estimate)
+
+        error_estimate = abs(np.sum(
+            error_weights * f(error_nodes),
+            axis=-1
+        ))
+
+        # print(error_estimate)
+
+        # TODO: not yet worrying about rounding error
+
+        return integral_estimate, error_estimate
+
+
+# For now, assuming that the underlying quadrature rules all have linear error
+# estimators so that we can combine the error nodes and error weights
+class ProductRule(CubatureRuleLinearError):
+    def __init__(self, base_rules):
+        self.nodes = _cartesian_product([rule.nodes for rule in base_rules])
+        self.error_nodes = _cartesian_product([rule.error_nodes for rule in base_rules])
+
+        self.weights = np.prod(
+            _cartesian_product([rule.weights for rule in base_rules]),
+            axis=0
+        )
+        self.error_weights = np.prod(
+            _cartesian_product([rule.error_weights for rule in base_rules]),
+            axis=0
+        )
+
```

Comment:
Rather then computing the `weights` and `error_weights` in the constructor, I think we want to make effective use of `functools.cached_property`.